HorizontalNavTab Extension

alecmerdler opened the pull request:
### Description

Implements the `DetailViewTab` extension type, which allows adding a new tab to an existing resource's detail view.

### Screenshots

**`Pods` detail view with demo plugin:**
![Screenshot_20200109_164724](https://user-images.githubusercontent.com/11700385/72116637-be5c7100-32ff-11ea-9963-0b70937ff0bc.png)

Needed for https://issues.redhat.com/browse/PROJQUAY-119
Blocked by https://github.com/openshift/console/pull/3383

christianvogt commented on frontend/public/components/utils/horizontal-nav.tsx:
```diff
@@ -201,7 +202,16 @@ export const HorizontalNav: React.FC<HorizontalNavProps> = React.memo((props) =>
   );
-  const pages = props.pages || props.pagesFor(_.get(props.obj, 'data'));
+  const pages = (props.pages || props.pagesFor(_.get(props.obj, 'data')))
+    // FIXME(alecmerdler): Handle `flags` for extension tabs...
```
This can more easily be handled later by using the hoc or hook from https://github.com/openshift/console/pull/3383

christianvogt commented on frontend/packages/console-plugin-sdk/src/typings/details.ts:
```diff
@@ -0,0 +1,23 @@
+  export interface DetailViewTab {
+    model: K8sKind;
+    flag?: string;
+    href: string;
```
What about path support?

vojtechszocs replied:
This can be done by reusing the `Page` type, see my comment below.

christianvogt commented on frontend/packages/console-plugin-sdk/src/typings/details.ts:
```diff
@@ -0,0 +1,23 @@
+namespace ExtensionProperties {
+  export interface DetailViewTab {
+    model: K8sKind;
+    flag?: string;
```
Just wait for https://github.com/openshift/console/pull/3383

alecmerdler replied:
Can I help with review or testing to get that pull request merged?

christianvogt replied:
@alecmerdler by all means.
There's just one pending typing fix left on my side for it.

christianvogt commented on frontend/packages/console-plugin-sdk/src/typings/details.ts:
```diff
@@ -0,0 +1,23 @@
+    href: string;
+    name: string;
+    // TODO(alecmerdler): Should we allow positioning of the tab...?
+    position?: number;
```
Should make it consistent with how we do other extensions like nav items.
But I prefer `weight: number` or using a relative position like `after: string`. Where the `after` value is an id (preferably) or a label.

alecmerdler replied:
We should just remove it for now, and add it once there is a legitimate use-case.

vojtechszocs commented on frontend/packages/console-plugin-sdk/src/typings/details.ts:
```diff
@@ -0,0 +1,23 @@
+import { K8sKind } from '@console/internal/module/k8s';
+import { PageComponentProps } from '@console/internal/components/utils/horizontal-nav';
```
@alecmerdler This is effectively about extending the common `HorizontalNav` component which is used on many places. If I read the code correctly, this covers more than just resource details pages.

I'd suggest to give this extension a more accurate name. For example, `HorizontalNavTab` as extension type name and `HorizontalNav/Tab` as `type` property value.

In general, the `type` value should reflect the context of the extension. For example, `Dashboards/Tab` means a tab on dashboards page, and `Dashboards/Overview/Inventory/Item` means new item in inventory which is part of overview on dashboards page.

vojtechszocs commented on frontend/public/components/factory/details.tsx:
```diff
@@ -63,6 +63,7 @@ export const DetailsPage = withFallback<DetailsPageProps>((props) => {
 
 export type Page = {
```
We could reuse this type (using TS `Pick` mapped type to represent a sub-structure) in extension properties declaration.

alecmerdler commented on frontend/public/components/utils/horizontal-nav.tsx:
```diff
@@ -7,7 +7,8 @@ import { Route, Switch, Link, withRouter, match, matchPath } from 'react-router-
 import { PodsPage } from '../pod';
 import { AsyncComponent } from './async';
-import { K8sResourceKind } from '../../module/k8s';
+import { K8sResourceKind, referenceForModel, referenceFor } from '../../module/k8s';
```
Running into circular dependency issue from these imports...

alecmerdler commented on frontend/packages/console-demo-plugin/src/plugin.tsx:
```diff
@@ -277,6 +280,21 @@ const plugin: Plugin<ConsumedExtensions> = [
     },
   },
+  {
+    type: 'HorizontalNavTab',
```
**NOTE:** It appears that the `@console/demo-plugin` is broken (circular dependency issue). But I've tested this plugin using the planned feature addition of https://github.com/openshift/console/pull/3874.

vojtechszocs replied:
Thanks for the heads-up, I will try to look into it.

Regardless, thanks @alecmerdler for keeping the demo plugin in mind when adding new extension type.

alecmerdler commented:
@vojtechszocs @christianvogt Rebased and using the new `useExtensions` React hook. I believe I addressed all comments. Please take a look when you have a chance.

christianvogt commented on frontend/public/components/utils/horizontal-nav.tsx:
```diff
@@ -201,7 +203,21 @@ export const HorizontalNav: React.FC<HorizontalNavProps> = React.memo((props) =>
     {},
   );
-  const pages = props.pages || props.pagesFor(_.get(props.obj, 'data'));
+  const navTabExtensions = _.get(props.obj, 'data')
```
we should use optional chaining and avoid lodash `get` now.

vojtechszocs replied:
Right, we also did this in #3907 - it's a great way to eliminate those `_.get` calls.

```ts
const navTabExtensions = props.obj?.data
```

christianvogt commented on frontend/public/components/utils/horizontal-nav.tsx:
```diff
@@ -201,7 +203,21 @@ export const HorizontalNav: React.FC<HorizontalNavProps> = React.memo((props) =>
   );
-  const pages = props.pages || props.pagesFor(_.get(props.obj, 'data'));
+  const navTabExtensions = _.get(props.obj, 'data')
+    ? useExtensions<HorizontalNavTab>(isHorizontalNavTab).filter(
```
Hooks shouldn't be used conditionally. Probably no harm if it's the only one in the function though.

vojtechszocs replied:
You can call the `useExtensions` hook multiple times.

It's better if a hook is called consistently (i.e. always once, instead of maybe once, per component render).

vojtechszocs commented on frontend/public/components/utils/horizontal-nav.tsx:
```diff
@@ -201,7 +203,21 @@ export const HorizontalNav: React.FC<HorizontalNavProps> = React.memo((props) =>
+    ? useExtensions<HorizontalNavTab>(isHorizontalNavTab).filter(
+        (tab) => referenceForModel(tab.properties.model) === referenceFor(props.obj.data),
+      )
+    : [];
```
Suggestion:

```ts
const navTabExtensions = useExtensions<HorizontalNavTab>(isHorizontalNavTab).filter(
  (tab) => props.obj?.data && referenceForModel(tab.properties.model) === referenceFor(props.obj.data),
);

const pages = (props.pages || props.pagesFor(props.obj?.data)).concat(
  navTabExtensions.map((tab) => ({
    ...tab.properties.page,
    component: (params: PageComponentProps) => (
      <AsyncComponent {...params} loader={tab.properties.loader} />
    ),
  })),
);
```

alecmerdler replied:
Done

alecmerdler commented:
/retest

vojtechszocs commented:
/lgtm

spadgett commented:
/hold
for merge queue fix #4065

spadgett commented:
/hold cancel
/retest

alecmerdler commented:
/hold

alecmerdler commented:
/hold cancel

alecmerdler commented:
/retest

alecmerdler commented:
/retest

alecmerdler commented:
/retest

alecmerdler commented:
/retest
